guide_drafts/extending-diesel.md

```diff
@@ -0,0 +1,385 @@
+Extending Diesel's Query Builder
+================================
+
+Diesel provides a lot of capabilities out of the box.
+However, it doesn't necessarily provide everything your app may want to use.
+One of Diesel's greatest strengths,
+is that it can be extended to add new functionality.
+
+In this guide we'll look at several ways to hook into Diesel's query builder,
+both to add new capabilities,
+and to introduce new abstractions.
+
+This guide is only going to cover extending the query builder.
+How to add support for new SQL types will be covered in a future guide.
+
+`sql_function!`
+---------------
+
+The easiest and most common way to extend Diesel's query builder
+is by declaring a new SQL function.
+This can be used for functions defined by your database,
+or for built-in functions that Diesel doesn't support out of the box.
+
+Functions in SQL often have multiple signatures,
+making them difficult or impossible to represent in Rust.
+Because of this, Diesel only provides support for a small number
+of built-in SQL functions.
+Consider `COALESCE`.
+This function can take any number of arguments,
+and its return type changes based on whether any arguments are `NOT NULL`.
+While we can't easily represent that in Rust,
+we can use `sql_function!` to declare it with the exact signature we're using.
+
+```
+use diesel::types::{Nullable, Text};
+sql_function!(coalesce, Coalesce, (x: Nullable<Text>, y: Text) -> Text);
+
+users.select(coalesce(hair_color, "blue"))
+```
+
+As this example shows,
+`sql_function!` converts it's argument like other parts of the query builder.
+This means that the generated function can take both Diesel expressions,
+and Rust values to be sent with the query.
+
+The macro takes three arguments:
+
+- A function name
+- A type name
+- A type signature
+
+The type signature uses the same syntax as a normal Rust function.
+However, the types given are SQL types,
+not concrete Rust types.
+This is what allows us to pass both columns and Rust strings.
+If we defined this function manually, it would look like this:
+
+```
+fn coalesce<X, Y>(x: X, y: Y) -> Coalesce<X::Expression, Y::Expression>
+where
+    X: AsExpression<Nullable<Text>>,
+    Y: AsExpression<Text>,
+```
+
+The type name given as the second argument is almost never used.
+Instead, a helper type is generated with the same name as the function.
+This helper type handle's Diesel's argument conversion.
+This lets us write `coalesce<hair_color, &str>`
+instead of `Coalsece<hair_color, Bound<Text, &str>>`.
+
+Custom SQL
```

Comment:
I would love an idea for a better header to use here.

Reply:
I originally had this as the third section (swapped with "custom operators"), since this is definitely "lower level" than custom operators are.

However, this seemed like code an app is more likely to write. Operators are more likely to be handled by libraries (and that example might even belong in the "supporting custom SQL types" guide).

Anyway, this order is roughly "most likely to be done by apps to least likely". If we swap this with "custom operators", the order becomes "highest level to lowest level"

Reply:
How about "Using custom SQL and how to extend the Query DSL"

I like that this takes such a prominent position in the guide; it's very useful information and more people should make use of this to build cool extension crates!